### ISSUE TITLE
Reuse of tjinstance (TJINIT_COMPRESS) has a bug when switching from lossless state back to lossy (setCompDefaults bug)

### ISSUE THREAD

**Issue by dzeitlin** [libjpeg-turbo/libjpeg-turbo#792]
Our code reuses tjinstance in compression mode (so called, pool of tjinstnaces), very typical usage on image-server.
Here is the simplest code to repoduce the bug.
```
tjhandle hcomp = tj3Init(TJINIT_COMPRESS);
tj3Set(hcomp, TJPARAM_LOSSLESS, 1); // switch to lossless mode
tj3Compress8(hcomp, ..., pBuffer); // pBuffer contains LOSSLESS data (ALL OK)

tj3Set(hcomp, TJPARAM_LOSSLESS, 0); // now lets switch back to LOSSY mode on the same instance
tj3Set(hcomp, TJPARAM_QUALITY, 90);
tj3Compress8(hcomp, ..., pBuffer) // <--- bug is here, still does lossless compression
```
the bug is in `setCompDefaults(tjinstance *this, int pixelFormat)` method
```
if (this->lossless) {
#ifdef C_LOSSLESS_SUPPORTED
    jpeg_enable_lossless(&this->cinfo, this->losslessPSV, this->losslessPt);
#endif
  ..
}
```
`else` is missed here and must be added to **reset cinfo mode back to lossy**, something like that BUT additional params are needed to be reset too
`cinfo->master->lossless = FALSE;`

But I think better to pass `this->lossless` state to jpeg_enable_lossless() function and it will support both states and switching between them on the same instance and set `cinfo` accordingly.
Thanks

**Comment by dzeitlin** [comment]
Looks it's more complex I think reseting `cinfo->master->lossless = FALSE;` didn't help for all qualities TJPARAM_QUALITY (somehow, very wierd, it helped for quality=80 but not for quality=90)
I think `jinit_c_master_control` function also have to be fixed.

Again the bug, switihing from LOSSY back to LOSSLESS on the *same tjinstance* behaves BAD.

**Comment by dcommander** [comment]
I have been without Internet access in my home office/lab for three weeks since Hurricane Helene, so please be patient. I will give this high priority as soon as I can return to normal business operations (hopefully next week.)

**Comment by dcommander** [comment]
Oof.  Yeah, this is really thorny, and I'm not sure yet how to solve it.  It affects both the libjpeg and TurboJPEG APIs.  The surface issue is that the libjpeg API provides no means of disabling lossless mode once it is enabled.  I tried modifying `jpeg_set_defaults()` to achieve that end, but then I ran into another issue.  The second issue is that the lossless codec allocates a different Huffman DC table than the lossy codec.  Huffman tables are allocated on the permanent pool, so they can't be freed until the libjpeg object is destroyed.  Thus, attempting to replace the lossless codec's Huffman DC table with a different Huffman DC table (a baseline Huffman DC table, for example) would leak memory.  The library instead just reuses the lossless codec's DC table, which produces a correct lossy JPEG image but not a baseline lossy JPEG image.  I will look at it more tomorrow.

**Comment by dzeitlin** [comment]
Currently we reuse the tjinstnace only of lossy mode (it's ~10% speedup relative to compression itself of 512*512 image tile).
Since lossless mode path is very slow and not optimized yet (x10-15 times slower), we don't cache it and create tjinstnace on every call and free it.

I suggest to ASSERT simply switch back to lossy mode back from lossless after first lossless performed COMPRESSION
[OR]
Another simple solution is to add TJINIT_COMPRESS_LOSSLESS
thus it can't be changed and must be specified at create level only once.

**Comment by dcommander** [comment]
@dzeitlin Yes, using a different TurboJPEG instance for lossy and lossless mode is an appropriate workaround for this at the moment.  Lossless JPEG was only implemented in libjpeg-turbo because MathWorks paid for it, and it is only really necessary because some variants of camera raw, DNG, and the DICOM medical imaging format use it.  Those formats only use it because they predate the existence of PNG.  Lossless JPEG is nice to have because libjpeg-turbo is an ISO/ITU-T reference implementation of JPEG-1, and the JPEG-1 spec includes a lossless variant.  Thus, all software that uses libjpeg-turbo (image viewers, etc.) immediately gained the ability to handle lossless JPEG-1 images by upgrading their version of libjpeg-turbo.  However, lossless JPEG-1 isn't a very useful format these days, and I doubt that anyone cares enough about its performance to pay me to optimize it.  To put some numbers on that, I did a [study](https://libjpeg-turbo.org/About/SmartScale-Lossless) years ago that compared the DCT-based lossless JPEG trick used in jpeg-8 and jpeg-9 (which was never accepted by the standards committees) with JPEG-LS, JPEG XR, webp, and PNG.  The same **artificial.ppm**, **nightshot_iso_100.ppm**, and **vgl_6434_0018.ppm** images perform as follows with lossless JPEG-1 on the same machine:

| artificial.ppm | Comp. ratio | Time (s) |
| ----- | ----- | ----- |
| Lossless JPEG-1 (libjpeg-turbo 3.0.3) | 4.924 | 0.355 |

| nightshot_iso_100.ppm | Comp. ratio | Time (s) |
| ----- | ----- | ----- |
| Lossless JPEG-1 (libjpeg-turbo 3.0.3) | 2.649 | 0.846 |

| vgl_6434_0018.ppm | Comp. ratio | Time (s) |
| ----- | ----- | ----- |
| Lossless JPEG-1 (libjpeg-turbo 3.0.3) | 4.839 | 0.069 |

Lossless JPEG-1 (as implemented in libjpeg-turbo) is actually faster than the other lossless formats, but its compression ratio is not impressive at all.  If someone wanted a fast lossless format, then optimizing libpng would be a much better use of time and money.

API changes are hard.  If I can't figure out how to reuse the same instance for both lossy and lossless compression, then the path of least resistance would be to simply throw an error if the caller tries to set `TJPARAM_LOSSLESS` to 0 after compression has already occurred with it set to 1, and document that behavior.  The libjpeg API already effectively disallows reusing the same instance for lossy and lossless, because there is no function to disable lossless mode after `jpeg_enable_lossless()` has been called.

**Comment by dcommander** [comment]
Should be fixed now
